File: apps/api/app/schemas/statement.py
"""
Pydantic schemas for statement upload and processing.
"""

from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field


class StatementProcessingStatus(str, Enum):
    """
    Processing lifecycle statuses for an uploaded statement document.
    """
    RECEIVED = "RECEIVED"
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class StatementUploadResponse(BaseModel):
    """
    Response returned when a bank statement is successfully received by the backend.
    """
    upload_id: str = Field(..., description="Unique statement upload tracking ID")
    user_id: int = Field(..., description="Authenticated user ID who uploaded the statement")
    filename: str = Field(..., description="Original filename of the uploaded statement")
    content_type: str = Field(..., description="MIME type of the uploaded statement")
    size_bytes: int = Field(..., description="File size in bytes")
    status: str = Field("RECEIVED", description="Upload processing status (e.g., RECEIVED)")
    message: str = Field(..., description="Human-readable status summary")
    uploaded_at: str = Field(..., description="ISO 8601 UTC timestamp of upload")


class StatementStatusResponse(BaseModel):
    """
    Response returned when querying statement processing status.
    """
    upload_id: str = Field(..., description="Unique statement upload tracking ID")
    user_id: int = Field(..., description="Authenticated user ID who owns the statement")
    filename: str = Field(..., description="Original filename of the statement")
    status: StatementProcessingStatus = Field(..., description="Current statement processing status")
    message: str = Field(..., description="Human-readable processing status description")
    created_at: str = Field(..., description="ISO 8601 UTC timestamp of initial upload")
    updated_at: str = Field(..., description="ISO 8601 UTC timestamp of last status update")
    error_detail: Optional[str] = Field(None, description="Error detail if processing failed")


class StatementStatusUpdateRequest(BaseModel):
    """
    Request model for transitioning statement processing status.
    """
    status: StatementProcessingStatus = Field(..., description="Target processing status")
    message: Optional[str] = Field(None, description="Optional custom status message")
    error_detail: Optional[str] = Field(None, description="Optional error details if status is FAILED")


class StatementPageText(BaseModel):
    """
    Extracted text content from an individual page of a statement document.
    Preserves page boundaries and position.
    """
    page_number: int = Field(..., description="1-indexed page number within the document")
    text: str = Field(..., description="Raw text extracted from this page")
    char_count: int = Field(..., description="Total characters extracted from this page")


class StatementDocumentMetadata(BaseModel):
    """
    Technical and structure metadata for an extracted statement.
    Stores format, page count, and character totals needed for downstream parsing.
    """
    format: str = Field(..., description="Document format: pdf or image type (e.g. image/png)")
    page_count: int = Field(..., description="Total pages/images in the document")
    total_characters: int = Field(..., description="Total characters extracted across all pages")
    extracted_at: str = Field(..., description="ISO 8601 UTC timestamp of extraction")
    file_size_bytes: int = Field(..., description="Original file size in bytes")


class StatementExtractionResponse(BaseModel):
    """
    Result returned by statement text-extraction / OCR processing.
    Contains raw text, per-page structured text, and document metadata.
    """
    upload_id: str = Field(..., description="Unique statement upload tracking ID")
    status: StatementProcessingStatus = Field(..., description="Resulting statement processing status")
    raw_text: str = Field("", description="Aggregated raw document text preserving page boundaries")
    pages: list[StatementPageText] = Field(default_factory=list, description="Per-page extracted text items")
    total_pages: int = Field(0, description="Total number of pages processed")
    total_characters: int = Field(0, description="Total number of characters extracted")
    document_metadata: Optional[StatementDocumentMetadata] = Field(
        None, description="Document structure and extraction metadata"
    )
    message: str = Field(..., description="Human-readable extraction status message")
    error_detail: Optional[str] = Field(None, description="Detailed error code/message if extraction failed")


class StatementOCRStatusResponse(BaseModel):
    """
    Health and availability diagnostics for the local statement OCR engine.
    """
    available: bool = Field(..., description="Whether a local OCR binary is configured and ready")
    engine: str = Field("tesseract", description="Primary OCR engine name")
    version: Optional[str] = Field(None, description="Installed OCR engine version if available")
    binary_path: Optional[str] = Field(None, description="Resolved path to OCR binary if found")
    supported_formats: list[str] = Field(
        default_factory=lambda: [
            "image/png",
            "image/jpeg",
            "image/jpg",
            "image/webp",
            "image/heic",
            "image/heif",
            "scanned_pdf",
        ],
        description="Supported document formats for optical character recognition",
    )
    message: str = Field(..., description="Availability status summary and operational guidance")
    install_instructions: Optional[dict[str, str]] = Field(
        None, description="Installation instructions if binary is not installed"
    )

File: apps/mobile/src/types/statement.ts
/**
 * Statement Types & Models
 * (`apps/mobile/src/types/statement.ts`)
 *
 * Strongly-typed models for bank/UPI statement document selection,
 * metadata validation, and backend upload responses.
 */

/**
 * Supported statement file categories.
 */
export type StatementFileType = "pdf" | "image";

/**
 * Validated metadata for a selected statement file.
 */
export interface SelectedStatementFile {
  /** Local file://, content://, or blob URI to the selected document */
  uri: string;
  /** Original filename including extension */
  name: string;
  /** File size in bytes */
  size: number;
  /** Resolved or normalized MIME type */
  mimeType: string;
  /** Categorized file type ('pdf' or 'image') */
  fileType: StatementFileType;
  /** Lowercase file extension including dot (e.g. '.pdf') */
  extension: string;
  /** Unix timestamp in milliseconds of last modification, if available */
  lastModified?: number;
}

/**
 * Typed parameters for uploading a statement file to the backend.
 */
export interface StatementUploadRequest {
  file: SelectedStatementFile;
}

/**
 * Response model returned by POST /api/v1/users/{userId}/statement/upload
 * (the personalized transaction-pattern engine's endpoint — see
 * apps/api/app/services/statement_parser_service.py::ingest_statement,
 * which now backs StatementUploadCard.tsx instead of the original
 * /api/v1/statements/upload OCR-preview endpoint).
 *
 * `upload_id`/`status`/`message`/`filename`/`size_bytes`/`uploaded_at`
 * are a compatibility superset so this existing UI component keeps
 * working unmodified — see that function's docstring. `status` is
 * always terminal (COMPLETED/FAILED): this call is synchronous
 * end-to-end, there is no RECEIVED/PROCESSING phase to poll for.
 */
export interface StatementUploadResponse {
  /** Unique statement upload tracking receipt ID (e.g., 'stmt_upl_...') */
  upload_id: string;
  /** Authenticated user ID */
  user_id: number;
  /** Original filename stored and acknowledged by backend */
  filename: string;
  /** MIME type verified by backend */
  content_type: string;
  /** Verified size in bytes */
  size_bytes: number;
  /** Upload status ('RECEIVED') */
  status: string;
  /** Human-readable status confirmation message */
  message: string;
  /** ISO 8601 UTC timestamp of upload */
  uploaded_at: string;

  // -- Personalized transaction-pattern engine fields (new) -----------
  /** How many transaction rows were successfully parsed from the file. */
  parsed_rows?: number;
  /** How many of those parsed rows were new (not already stored). */
  inserted_rows?: number;
  /** How many were already-seen duplicates, skipped. */
  duplicate_rows?: number;
  /** "table" (pdfplumber found a structured table) or "ocr_fallback"
   *  (text/OCR extraction + line-based parsing — see
   *  statement_parser_service.py's OCR FALLBACK docstring note). */
  parse_method?: "table" | "ocr_fallback";
  /** Whether this upload triggered an immediate bootstrap training run
   *  (a user's very first statement upload — see
   *  user_pattern_trainer.py). */
  trained?: boolean;
  /** Personalized baseline percentiles, once trained (null until enough
   *  history exists). */
  p50_amount?: number | null;
  p90_amount?: number | null;
  p99_amount?: number | null;
}

/**
 * Classification of statement upload errors.
 */
export type StatementUploadErrorType =
  | "AUTH_FAILURE"        // 401 Unauthorized / missing / expired token
  | "NETWORK_FAILURE"     // Network disconnect / timeout / server unreachable
  | "SERVER_VALIDATION"   // 400 Bad Request / unsupported format / empty or oversized file
  | "SERVER_ERROR"        // 500+ Internal server error
  | "CANCELLED"           // Upload cancelled before or during flight
  | "UNKNOWN_ERROR";      // Uncategorized runtime failure

/**
 * Strongly typed error details for statement upload failures.
 */
export interface StatementUploadError {
  type: StatementUploadErrorType;
  message: string;
  statusCode?: number;
  details?: any;
}

/**
 * Discriminated union outcome of a statement upload operation.
 */
export type StatementUploadResult =
  | { success: true; data: StatementUploadResponse }
  | { success: false; error: StatementUploadError };

/**
 * Processing lifecycle statuses for an uploaded statement.
 */
export type StatementProcessingStatus =
  | "RECEIVED"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED";

/**
 * Response model returned by GET /api/v1/statements/{upload_id}/status.
 */
export interface StatementStatusResponse {
  /** Unique statement upload tracking receipt ID */
  upload_id: string;
  /** Authenticated user ID who owns the statement */
  user_id: number;
  /** Original filename of the statement */
  filename: string;
  /** Current processing status */
  status: StatementProcessingStatus;
  /** Human-readable status description */
  message: string;
  /** ISO 8601 UTC timestamp of initial upload */
  created_at: string;
  /** ISO 8601 UTC timestamp of last status transition */
  updated_at: string;
  /** Error detail if status is FAILED */
  error_detail?: string | null;
}

/**
 * Discriminated union outcome of a statement processing status query.
 */
export type StatementStatusResult =
  | { success: true; data: StatementStatusResponse }
  | { success: false; error: StatementUploadError };

/**
 * Text content extracted from an individual page of a bank statement.
 */
export interface StatementPageText {
  /** 1-indexed page number within the document */
  page_number: number;
  /** Raw text extracted from this page */
  text: string;
  /** Total characters extracted from this page */
  char_count: number;
}

/**
 * Technical and structural metadata for an extracted document.
 */
export interface StatementDocumentMetadata {
  /** Document MIME or format type */
  format: string;
  /** Total pages or images in the document */
  page_count: number;
  /** Total characters extracted across all pages */
  total_characters: number;
  /** ISO 8601 UTC timestamp of extraction */
  extracted_at: string;
  /** Original file size in bytes */
  file_size_bytes: number;
}

/**
 * Response model returned by POST /api/v1/statements/{upload_id}/extract
 * and GET /api/v1/statements/{upload_id}/extraction.
 */
export interface StatementExtractionResponse {
  /** Unique statement upload tracking receipt ID */
  upload_id: string;
  /** Processing status (COMPLETED or FAILED) */
  status: StatementProcessingStatus;
  /** Aggregated raw document text preserving page boundaries */
  raw_text: string;
  /** Per-page extracted text chunks */
  pages: StatementPageText[];
  /** Total number of pages processed */
  total_pages: number;
  /** Total number of characters extracted */
  total_characters: number;
  /** Document metadata */
  document_metadata?: StatementDocumentMetadata | null;
  /** Human-readable extraction summary */
  message: string;
  /** Error details if extraction failed */
  error_detail?: string | null;
}

/**
 * Discriminated union outcome of a statement extraction trigger or query.
 */
export type StatementExtractionResult =
  | { success: true; data: StatementExtractionResponse }
  | { success: false; error: StatementUploadError };

/**
 * Health and availability diagnostics for the backend statement OCR engine.
 */
export interface StatementOCRStatusResponse {
  /** Whether a local OCR engine binary is available */
  available: boolean;
  /** Primary OCR engine name */
  engine: string;
  /** Resolved OCR engine version */
  version?: string | null;
  /** Resolved binary path on backend host */
  binary_path?: string | null;
  /** Supported document formats */
  supported_formats: string[];
  /** Summary message and operational instructions */
  message: string;
  /** Platform install instructions if unavailable */
  install_instructions?: Record<string, string> | null;
}

/**
 * Discriminated union outcome of an OCR engine availability query.
 */
export type StatementOCRStatusResult =
  | { success: true; data: StatementOCRStatusResponse }
  | { success: false; error: StatementUploadError };
